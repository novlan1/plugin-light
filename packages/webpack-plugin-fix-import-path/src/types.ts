export interface IFixImportPathOptions {
  // 待处理的文件列表
  handleList?: Array<string>;
  // 插件名称
  pluginName?: string;
}
