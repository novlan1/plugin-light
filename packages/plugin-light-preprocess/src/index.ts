export { preprocessFile, preprocessFileSync, preprocess } from './preprocess';
