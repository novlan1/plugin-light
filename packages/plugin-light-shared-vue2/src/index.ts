export {
  shouldUseLoader,
} from './loader-options';


export {
  parseSFC,
  composeSFC,
} from './sfc';

