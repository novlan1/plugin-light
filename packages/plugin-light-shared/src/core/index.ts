export * from './cross-game-style';
export * from './gen-version';
export * from './ifdef';
export * from './inject-dynamic-style';
export * from './replace-vue-directive';
export * from './v-lazy';
