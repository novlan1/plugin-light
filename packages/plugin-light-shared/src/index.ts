export * from './core';


export {
  checkBundleAnalyze,
  checkDebugMode,
} from './bundle-analyze';

export {
  PLATFORM_MAP,
  ALL_PLATFORM,
  PLATFORMS_ALL,
  PLATFORMS_MP,
  HTML_MAP,
  CSS_MAP,
  CSS_POSTFIX_MAP,
  CDN_MAP,
} from './config';

export {
  DEFAULT_TRANSPILE_DEPENDENCIES,
  DEFAULT_ADAPTER_DIRS,
  AEGIS_EXTERNAL_SCRIPT_LINK,
  UNI_SIMPLE_ROUTER_SCRIPT_LINK,
  EXTERNAL_LINK_MAP,
} from './config-project';

export {
  ROOT_NAME,
  saveJsonToLog,
  createLogDir,
  normalizePath,
  updateAssetSource,
  removeFirstSlash,
  parseSetDeps,
  getRelativePath,
  getUniCliCache,
} from './helper';


export {
  DEFAULT_CONTEXT_OBJECT,
} from './context';


export {
  getDeps,
} from './deps';


export {
  findDependencies,
} from './find-dependencies';


export {
  checkH5,
} from './h5';


export {
  getLoaderFile,
  getLoaderProdFile,
} from './loader-file';

export {
  saveLoaderLog,
  recordLoaderLog,
} from './loader-log';

export {
  getProjectName,
  getSubProjectName,
} from './project-name';

export {
  updateManifest,
  revertManifest,
} from './replace-manifest';

export {
  getRootDir,
} from './root';


export {
  getSubProjectConfig,
  getSubProjectRoot,
} from './sub-project';
