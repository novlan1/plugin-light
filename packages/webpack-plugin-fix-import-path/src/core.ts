import path from 'path';


/**
 * 像 require('../common/runtime.js'); 这种引用路径是有问题的，
 * 应该是 ../../../common/runtime.js
 *
 */
export function fixImportPath({
  source,
  filePath,
  root,
  assets,
  handleList,
}: {
  source: string;
  filePath: string;
  root: string;
  assets: Record<string, any>;
  handleList: Array<string>;
}) {
  if (!handleList.length) {
    return {
      newSource: source,
    };
  }


  const reg = new RegExp(`require\\((?:'|")([./]*)(${handleList.join('|')})(?:'|")\\)`, 'g');
  const logData = [];

  let match = reg.exec(source);
  if (!match) {
    return {
      newSource: source,
    };
  }

  let newSource = source;


  while (match) {
    const { 0: origin, 1: relative, 2: innerFile } = match;

    let curPath = path.resolve(filePath, relative, '../');
    if (!relative) {
      // 绝对路径
      curPath = path.dirname(curPath);
    }

    const curTargetRelative = getCurTargetRelative({
      relative,
      innerFile,
      filePath,
    });

    if (!curTargetRelative || !assets[curTargetRelative]) {
      const targetFile = findRealFile({
        filePath: curPath,
        targetFile: innerFile,
        root,
        assets,
      });

      if (targetFile) {
        let relative = path.relative(path.dirname(filePath), targetFile);

        if (relative) {
          if (!relative.startsWith('.')) {
            relative = `./${relative}`;
          }

          const target = `require('${relative}')`;
          newSource = newSource.replace(origin, target);

          logData.push({
            origin,
            target,
            targetFile,
          });
        }
      }
    } else {
      logData.push({
        origin,
        hold: true,
        curTargetRelative,
      });
    }

    match = reg.exec(newSource);
  }


  return { newSource, logData };
}


function getCurTargetRelative({
  relative,
  innerFile,
  filePath,
}: {
  relative: string;
  innerFile: string;
  filePath: string;
}) {
  let curTargetRelative = '';
  if (relative) {
    const curTarget = path.resolve(path.dirname(filePath), `${relative}${innerFile}`);
    curTargetRelative = path.relative(path.dirname(filePath), curTarget);
  }

  return curTargetRelative;
}


function findRealFile({
  filePath,
  targetFile,
  root,
  assets,
}: {
  filePath: string;
  targetFile: string;
  root: string;
  assets: Record<string, any>;
}): string {
  if (filePath.length < root.length) return '';

  const dir = path.dirname(filePath);
  const newFilePath = path.resolve(dir, targetFile);
  const newFileRelativePath = path.relative(root, newFilePath);

  if (assets[newFileRelativePath]) {
    return newFilePath;
  }

  return findRealFile({
    filePath: dir,
    targetFile,
    root,
    assets,
  });
}
