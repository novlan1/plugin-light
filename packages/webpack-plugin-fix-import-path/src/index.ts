export { FixImportPathPlugin } from './fix-import-path';
export * from './types';
