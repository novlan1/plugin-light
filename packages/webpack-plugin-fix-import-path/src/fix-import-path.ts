import path from 'path';
import { replaceAllPolyfill } from 't-comm';

import { fixImportPath } from './core';
import { createLogDir, saveJsonToLog,  updateAssetSource } from '@plugin-light/shared';
import type { IFixImportPathOptions } from './types';


replaceAllPolyfill();


const HANDLE_LIST = [
  'common/vendor.js',
  'common/runtime.js',
  'common/main.js',
];


export class FixImportPathPlugin {
  outputDir = process.env.UNI_OUTPUT_DIR || '';
  handleList: Array<string> = HANDLE_LIST;
  options?: IFixImportPathOptions;


  constructor(options: IFixImportPathOptions) {
    this.options = options;

    if (options?.pluginName) {
      this.outputDir = path.resolve(this.outputDir, options.pluginName);
    }
    if (options?.handleList) {
      this.handleList = options.handleList;
    }
    createLogDir();
  }

  apply(compiler: any) {
    if (!this.handleList.length) {
      return;
    }

    compiler.hooks.emit.tap('FixImportPathPlugin', (compilation: any) => {
      const { assets } = compilation;
      this.modifyRef(assets);
    });
  }


  modifyRef(assets: Record<string, any>) {
    const allLogData = [];

    for (const key of Object.keys(assets)) {
      const value = assets[key];
      if (key.endsWith('.js')) {
        const source = value.source().toString();

        const filePath = path.resolve(this.outputDir, key);
        const { newSource, logData = [] } = fixImportPath({
          source,
          filePath,
          root: this.outputDir,
          assets,
          handleList: this.handleList,
        });

        updateAssetSource(assets, key, newSource);

        allLogData.push({
          file: filePath,
          logData,
        });
      }
    }

    saveJsonToLog(allLogData, 'fix-import-path.result.json');
  }
}

