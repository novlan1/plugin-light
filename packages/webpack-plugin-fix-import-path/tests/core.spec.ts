import path from 'path';
import { fixImportPath } from '../src/core';


const assets = {
  'common/vendor.js': {},
  'views/index/common/vendor.js': {},
  'views/index/index.js': {},
  'views/setting/common/vendor.js': {},
  'views/setting/setting.js': {},
};

const outputDir = '/Users/user/dzs-match/dist/build/mp-weixin';

const defaultHandleList = [
  'common/vendor.js',
  'common/runtime.js',
  'common/main.js',
];


describe('fixImportPath', () => {
  const oldEnv = process.env;

  beforeEach(() => {
    process.env = { ...oldEnv };
  });

  afterEach(() => {
    process.env = oldEnv;
    jest.clearAllMocks();
  });


  it('should work correctly', () => {
    const source = 'require(\'../common/vendor.js\');';
    const filePath = path.resolve(outputDir, 'views/setting/setting.js');

    const { newSource, logData = [] } = fixImportPath({
      source, filePath, root: outputDir, assets,
      handleList: defaultHandleList,
    });

    expect(newSource).toBe('require(\'../../common/vendor.js\');');
    expect(logData).toMatchSnapshot();
  });

  it('should not modify anything if there is no match', () => {
    const source = 'require(\'../../views/search/common/vendor.js\');';
    const filePath = path.resolve(outputDir, 'views/setting/setting.js');

    const { newSource, logData = [] } = fixImportPath({
      source, filePath, root: outputDir, assets,
      handleList: defaultHandleList,
    });

    expect(newSource).toBe(source);
    expect(logData).toHaveLength(0);
  });

  it('should work correctly if there are multiple matches', () => {
    const source = 'require(\'./common/vendor.js\');require(\'../common/vendor.js\');';
    const filePath = path.resolve(outputDir, 'views/setting/setting.js');

    const { newSource, logData = [] } = fixImportPath({
      source, filePath, root: outputDir, assets,
      handleList: defaultHandleList,
    });

    expect(newSource).toBe('require(\'./common/vendor.js\');require(\'../../common/vendor.js\');');
    expect(logData).toMatchSnapshot();
  });
});
