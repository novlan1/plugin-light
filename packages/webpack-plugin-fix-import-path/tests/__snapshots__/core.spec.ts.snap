// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixImportPath should work correctly 1`] = `
[
  {
    "origin": "require('../common/vendor.js')",
    "target": "require('../../common/vendor.js')",
    "targetFile": "/Users/user/dzs-match/dist/build/mp-weixin/common/vendor.js",
  },
]
`;

exports[`fixImportPath should work correctly if there are multiple matches 1`] = `
[
  {
    "curTargetRelative": "common/vendor.js",
    "hold": true,
    "origin": "require('./common/vendor.js')",
  },
  {
    "origin": "require('../common/vendor.js')",
    "target": "require('../../common/vendor.js')",
    "targetFile": "/Users/user/dzs-match/dist/build/mp-weixin/common/vendor.js",
  },
]
`;
